pkg/ingester/flush.go

```diff
@@ -327,9 +328,11 @@ func (i *Ingester) flushChunks(ctx context.Context, fp model.Fingerprint, labelP
 		)
 
 		start := time.Now()
+		streamsMtx.Lock()
 		if err := c.Encode(); err != nil {
 			return err
 		}
+		streamsMtx.Unlock()
```

Comment:
I think you should do a defer here.

Reply:
Can you explain who is racing for the stream data here ? Querying ?

Reply:
I think we should not defer `Unlock`. The reason being putting chunks to store here would be locked and other flush threads won't be able to call `put` until the lock is released. This would become sequential `io` and eventually make all flush threads sequential.

https://github.com/grafana/loki/blob/e22f365025cf0dc091f343ebda75141ed733f2f5/pkg/ingester/flush.go#L337

Reply:
If you return an error you're deadlocked.

Reply:
> Can you explain who is racing for the stream data here ? Querying ?

Yeah, it is `querying` which access the stream chunks here
https://github.com/grafana/loki/blob/e22f365025cf0dc091f343ebda75141ed733f2f5/pkg/ingester/stream.go#L261

but this is Read protected with streamMtx lock here 
https://github.com/grafana/loki/blob/e22f365025cf0dc091f343ebda75141ed733f2f5/pkg/ingester/instance.go#L316

since `c.Encode` is not write locked it causes the race.

Also, reading first chunk and getting bounds on that here causes the race.
https://github.com/grafana/loki/blob/e22f365025cf0dc091f343ebda75141ed733f2f5/pkg/ingester/flush.go#L174

Reply:
firstTime, _ := stream.chunks[0].chunk.Bounds()  ? this one is fixed or already ok ?

Reply:
> If you return an error you're deadlocked

oh yeah. You are right. Didn't realize that.

Reply:
> firstTime, _ := stream.chunks[0].chunk.Bounds() ? this one is fixed or already ok ?

this was already protected

Reply:
Great !